### PR TITLE
fix metallb matchlabels to be valid

### LINKED ISSUE #247: metallb `nomatch` labels violate valid label spec
In the migration to [LoadBalancer](https://pkg.go.dev/k8s.io/cloud-provider#LoadBalancer)  spec that we did in #224 , we had to find some way for metallb to track what services are served by each upstream peer-local pairing. This allowed us to then update or remove a pairing (as expected by the cloud-provider [LoadBalancer](https://pkg.go.dev/k8s.io/cloud-provider#LoadBalancer) API.

However, metallb provides no way of tracking it, nor does it have any additional comments or unused fields in its control configmap. We eventually settled on using the `match-labels` part of the metallb configmap.

For example:

```yaml
- my-asn: 65000
  peer-asn: 65530
  peer-address: 169.254.255.1
  peer-port: 0
  source-address: 10.68.190.5
  hold-time: ""
  router-id: ""
  node-selectors:
  - match-labels:
      kubernetes.io/hostname: dc-worker-1
    match-expressions: []
  - match-labels:
      nomatch.metal.equinix.com/service-name: default/nginx-deployment
    match-expressions: []
```

Since `node-selectors` is logical OR, the label `nomatch.metal.equinix.com/service-name` never will match anything, and so will be ignored by metallb, while allowing us to use it to track services. It also prevents us from having to have a pairing entry for each service, which would make the configmap grow to an untenable size very quickly (and probably break metallb).

However, this is invalid:

```yaml
  - match-labels:
      nomatch.metal.equinix.com/service-name: default/nginx-deployment
```

If we read the [description of valid characters](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set), it says:

> Valid label value:
> 
> * must be 63 characters or less (can be empty),
> * unless empty, must begin and end with an alphanumeric character ([a-z0-9A-Z]),
> * could contain dashes (-), underscores (_), dots (.), and alphanumerics between.

That means that `nomatch.metal.equinix.com/service-name` is perfectly fine as a key, but `default/nginx-deployment` (representing the service `nginx-deployment` in the namespace `default`) is not.

This now is breaking metallb.

Thanks to @enkelprifti98 for finding this.

We need a solution.

Some possibilities:

* replace the `/` we conventionally use to separate namespace from service, with some other valid character. Unfortunately, almost all valid characters also might be part of the namespace or service. Would `default_nginx_deployment` represent the deployment `nginx_deployment` in `default`, or `deployment` in `nginx_default`?
* use the `.` to separate them, as in `default.nginx-deployment`. This only works if `.` is not valid in either namespace or service name. I'm not sure of that. The page for naming is [here](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/), but it isn't clear what constraints namespaces and services have.
* hash or base64 the whole thing. Since we always know what we are comparing against, any consistent forward hash/base64 should work. E.g. instead of `nomatch.metal.equinix.com/service-name: default/nginx-deployment` we could do `nomatch.metal.equinix.com/service-name-base64: ZGVmYXVsdC9uZ2lueC1kZXBsb3ltZW50`. It is ugly, but it is usable and predictable, and has precedent in kubernetes' secrets
* other ideas

Thoughts @displague @cprivite @detiber @ocobleseqx ?

### PR DESCRIPTION
**Pull request by deitch**
fixes #247 
cc @displague @enkelprifti98

### CODE REVIEW

**Comment by displague** [comment]
Let's add an example to https://github.com/equinix/cloud-provider-equinix-metal#metallb

> CCM itself does not deploy the load-balancer or any part of it, including the ConfigMap. It only modifies an existing ConfigMap. This can be deployed by the administrator separately, using the manifest provided in the releases page, or in any other manner.

Maybe something like this:

The CCM modifies the MetalLB ConfigMap with special node-selectors. These node-selectors are ignored by MetalLB but are critical to the CCM management of MetalLB configuration. For example (perhaps we can make this look more practical):
```
- my-asn: 65000
  peer-asn: 65530
  peer-address: 169.254.255.1
  peer-port: 0
  source-address: 10.68.190.5
  hold-time: ""
  router-id: ""
  node-selectors:
  - match-labels:
      kubernetes.io/hostname: dc-worker-1
  - match-labels:
      nomatch.metal.equinix.com/service-namespace: default
      nomatch.metal.equinix.com/service-name: nginx-deployment
 - match-labels:
      nomatch.metal.equinix.com/service-namespace: ai
      nomatch.metal.equinix.com/service-name: trainer
 - match-labels:
      nomatch.metal.equinix.com/service-namespace: database
      nomatch.metal.equinix.com/service-name: mysql1
```

File: metal/loadbalancers/metallb/metallb.go [hunk 1]
```diff
@@ -15,10 +15,11 @@ import (
 )
 
 const (
-	hostnameKey      = "kubernetes.io/hostname"
-	serviceNameKey   = "nomatch.metal.equinix.com/service-name"
-	defaultNamespace = "metallb-system"
-	defaultName      = "config"
+	hostnameKey         = "kubernetes.io/hostname"
+	serviceNameKey      = "nomatch.metal.equinix.com/service-name"
+	serviceNameSpaceKey = "nomatch.metal.equinix.com/service-namespace"
```
Comment (displague) [inline metal/loadbalancers/metallb/metallb.go:20]: Space is vast.

```suggestion
	serviceNamespaceKey = "nomatch.metal.equinix.com/service-namespace"
```

File: metal/loadbalancers/metallb/configmap.go [hunk 2]
```diff
@@ -484,51 +487,67 @@ func (p *Peer) EqualIgnoreService(o *Peer) bool {
...
-func (p *Peer) Services() []string {
+func (p *Peer) Services() []Resource {
+	var services []Resource
 	for _, ns := range p.NodeSelectors {
+		var name, namespace string
 		for k, v := range ns.MatchLabels {
-			if k == serviceNameKey {
-				return strings.Split(v, ",")
+			switch k {
+			case serviceNameKey:
```
Comment (displague) [inline metal/loadbalancers/metallb/configmap.go:496]: What are your thoughts on _healing_ the old format?  We would detect when `serviceNameKey` contains the old format and append any/all services detected.

I presume we would need to build a similar capability into the other for loops walking nodeselectors (which makes me wonder if these could be more consolidated).

File: metal/loadbalancers/metallb/metallb.go (hunk 1 above)
Comment (deitch) [reply metal/loadbalancers/metallb/metallb.go:20]: It is, indeed, the final frontier. We can do this.

File: metal/loadbalancers/metallb/configmap.go (hunk 2 above)
Comment (deitch) [reply metal/loadbalancers/metallb/configmap.go:496]: Is there any point? Given that the old format is broken, and shouldn't work, why bother?

**Comment by deitch** [comment]
Updated the README. Ready for another look @displague

**Review by displague — APPROVED** [review]

File: metal/loadbalancers/metallb/configmap.go (hunk 2 above)
Comment (displague) [reply metal/loadbalancers/metallb/configmap.go:496]: Fair.

**Pull request merged**